[IMP] 7.0 version of module stock_obsolete

mdietrichc2c opened the pull request:
This is a port of the 6.0 module "stock_obsolete", for 7.0.

mdietrichc2c committed:
[IMP] 7.0 version of module stock_obsolete

mdietrichc2c committed:
[FIX] code cleanup + unit test added

lepistone commented on stock_obsolete/__init__.py:
```diff
@@ -0,0 +1,32 @@
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
+import wizard
```
`from .`

lepistone commented on stock_obsolete/__openerp__.py:
```diff
@@ -0,0 +1,51 @@
+        stock product, with an amount of depreciation. Your finance can now
+        take in charge a factor for each product, based on the depreciation
+        value for each product. You can set the depreciation value based on the
+        out of stock amount for each product for last 12/24 month.
```
for the last ... months

lepistone commented on stock_obsolete/report/__init__.py:
```diff
@@ -0,0 +1,30 @@
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
+import product_obsolete
```
`from .`

lepistone commented on stock_obsolete/report/product_obsolete.py:
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python2.3
```
wow, python2.3! that whole header should probably go

lepistone commented on stock_obsolete/report/product_obsolete.py:
```diff
@@ -0,0 +1,76 @@
+        super(Product_obsolete, self).__init__(cr, uid, name, context)
+        self.localcontext.update({
+            'time': time,
+            # 'lines': self.lines,
```
if we don't need that, better delete the line outright.

lepistone commented on stock_obsolete/report/product_obsolete.py:
```diff
@@ -0,0 +1,76 @@
+            return self.comma_me(new)
+
+    def get_depreciation_name(self, value):
+        if value == 'no':
```
We can live with the hardcoded selection values (it could be found from the _columns attribute of the model, but, I'm not blocking over that. Still, that would not get translated. Could you add _('Half') ?

lepistone commented on stock_obsolete/stock_obsolete.py:
```diff
@@ -0,0 +1,209 @@
+                                            'child_of',
+                                            location_ids)],
+                                          context=context)
+            res = loc_obj._product_get_multi_location_for_period(cr,
```
can you please do something like

``` python
res = loc_obj._product_get_multi_location_for_period(
    cr, uid, localtion_ids, ...
)
```

lepistone commented on stock_obsolete/stock_obsolete.py:
```diff
@@ -0,0 +1,209 @@
+
+    # Considering all state for obsolescence
+    # May be only done...
+    _product_out_qty_till_12m = _get_product_obsolescence_func(('confirmed',
```
same here

lepistone commented on stock_obsolete/stock_obsolete.py:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    _defaults = {
+        'depreciation': lambda *a: 'no',
```
`'no'` is enough, withough `lambda`

lepistone commented on stock_obsolete/wizard/__init__.py:
```diff
@@ -0,0 +1,30 @@
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
+import wizard_product_obsolescence
```
`from . import`

lepistone commented:
I added some minor remarks, thanks @mdietrichc2c! Once that's sorted, I'm :+1:

mdietrichc2c commented:
@lepistone thanks for the feedback! I corrected according to your remarks, and also added a unit test for the basic behaviour.

lepistone commented:
thanks! :+1:

pedrobaeza commented on stock_obsolete/report/product_obsolete.py:
```diff
@@ -0,0 +1,95 @@
+from openerp.tools.translate import _
+
+
+class Product_obsolete(report_sxw.rml_parse):
```
Please use CamelCase notation, not a mixed one.

pedrobaeza commented on stock_obsolete/stock_obsolete.py:
```diff
@@ -0,0 +1,200 @@
+                                             day=1,
+                                             days=-1)).strftime('%Y-%m-%d')
+        # Construct sql clause
+        sql_clause = "and date_expected between '%s' and '%s' " \
```
Can you avoid SQL, or it's strictively needed?

pedrobaeza commented:
Functionally speaking, in my country (Spain), you cannot make this kind of devaluation in your stock valuation. You have to take it to inventory loss when fully degradated. Is this allowed in any country?

gurneyalex commented:
@mdietrichc2c can you look at pedro's comments?

mdietrichc2c commented:
@pedrobaeza I changed the case for the report.
RE: SQL, given that the location list is found by a regular search in the request, and that the request contains a sum() and groupbys, it could be done with OpenERP but would be much slower for 2 non-stored function fields.
RE: functionality, I ported it to 7.0 for one of our existing customers, but only for the outgoing fields; they don't use the deprecation (and it's not used in the function fields as well). Maybe a renaming + removing the obsolescence would be in order.

yvaucher commented:
@mdietrichc2c don't you want to squash your 3 commits for PEP8 in only one ?

mdietrichc2c commented:
@yvaucher done, thanks

yvaucher commented:
@mdietrichc2c thanks :+1:

max3903 commented on stock_obsolete/__openerp__.py:
```diff
@@ -0,0 +1,54 @@
+    "version": "1.0",
+    "author": "Camptocamp",
+    "category": "Generic Modules/Stock",
+    "module":
```
I think it should be "description" here and init_xml, demo_xml and update_xml are deprecated.
:-1:

mdietrichc2c replied:
@max3903 fixed now, thanks

mdietrichc2c committed:
Rename keys for 7.0

jgrandguillaume commented:
Hi Pedro,

I think we have two things here:

1) The way you get the information (about wich product is obsolete or not)
2) The way you create the accounting entries

This module give one way to get some informations about your stock, to know what products has been used or not since the last 12/24 months. From that report, we can then decide to create you entries (inventory loss) as you want.

The thing is here, may be every company has a different way of computing this. In our case, one of our customer use this one. It doesn't means it's THE generic way to do it. Other people may look at the expiry date, or whatever other criteria.

So my question here is:

**Would you see this module a generic enough to be committed here or should we committ it in our customer specific branch ?**

Thanks for sharing opinions.

Regards,

Joël

pedrobaeza commented:
Well, I have tried the module and it seems a little incomplete, because it doesn't offer a way to change product depreciation, so as you have said, there are two parts, but I think none of them is completely accomplished. For the first part (reporting), I have a module that fulfill this located in stock section: http://bazaar.launchpad.net/~pedro.baeza/serviciosbaeza-openerp-addons/7.0/files/head:/stock_products_no_moves/

yvaucher commented on stock_obsolete/__init__.py:
```diff
@@ -0,0 +1,32 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
```
Please change license to AGPL instead of GPL

yvaucher commented on stock_obsolete/__openerp__.py:
```diff
@@ -0,0 +1,53 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
```
We want AGPL

yvaucher commented on stock_obsolete/__openerp__.py:
```diff
@@ -0,0 +1,53 @@
+##############################################################################
+
+{
+    "name": "Stock obsolete",
```
Here is the right place to add a tiny `summary`

yvaucher commented on stock_obsolete/report/__init__.py:
```diff
@@ -0,0 +1,30 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
```
AGPL

yvaucher commented on stock_obsolete/report/product_obsolete.py:
```diff
@@ -0,0 +1,95 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
```
AGPL

yvaucher commented on stock_obsolete/report/product_obsolete.py:
```diff
@@ -0,0 +1,95 @@
+
+    def set_context(self, objects, data, ids, report_type=None):
+        prod_obj = self.pool.get('product.product')
+        if data.get('ids', False) and \
```
Not needed to use \ you can write your condition adding parenthesis

yvaucher commented on stock_obsolete/stock_obsolete.py:
```diff
@@ -0,0 +1,206 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
```
AGPL again

yvaucher commented on stock_obsolete/wizard/__init__.py:
```diff
@@ -0,0 +1,30 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
```
AGPL

yvaucher commented on stock_obsolete/wizard/wizard_product_obsolescence.py:
```diff
@@ -0,0 +1,96 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+#
+##############################################################################
```
AGPL

yvaucher commented on stock_obsolete/wizard/wizard_product_obsolescence.py:
```diff
@@ -0,0 +1,96 @@
+        result = mod_obj._get_id(cr, uid, 'stock', 'stock_location_company')
+        stock_location_id = mod_obj.read(cr, uid,
+                                         [result],
+                                         ['res_id'])[0]['res_id']
```
missing context

yvaucher commented on stock_obsolete/wizard/wizard_product_obsolescence.py:
```diff
@@ -0,0 +1,96 @@
+        mod_obj = self.pool.get('ir.model.data')
+        result = mod_obj._get_id(cr, uid, 'stock_obsolete',
+                                 'product_product_obsolet_tree_view')
+        tree_view_id = mod_obj.read(cr, uid, [result], ['res_id'])[0]['res_id']
```
missing context=context

gurneyalex commented:
@yvaucher we can't really change the licence like this...

yvaucher commented:
@gurneyalex Author is @jgrandguillaume so I assume we can under the consentement of Joël

jgrandguillaume commented:
It's ok for me to change the license here.

lepistone commented:
Hi @mdietrichc2c @pedrobaeza. I'd like to make a decision and close this PR.

I can't say about the functional domain, but maybe as a compromise we can merge with an additional note on the doc mentioning the limitations pointed by @pedrobaeza, and maybe with a link to his module? I vote :+1: to do so.

(note: as noted above, we need to set the license to AGPL)

pedrobaeza commented:
For me it's OK.

lepistone commented:
Thanks. @mdietrichc2c if you can add a little note in the readme as discussed and make sure the license is AGPL everywhere, I can merge.

mdietrichc2c committed:
Change licence to AGPL and add contexts to read() methods

mdietrichc2c commented:
@lepistone @pedrobaeza @yvaucher okay: I changed the licence in all files, added the context to the read() methods, and added a small description + summary to **openerp**.py

mdietrichc2c committed:
Add 'r' repfix to regular expressions

pedrobaeza commented:
One minor PEP8 error left.

pedrobaeza commented:
What about my comment about SQL?

mdietrichc2c committed:
PEP8 error

mdietrichc2c commented:
@pedrobaeza I responded back in October: 
"RE: SQL, given that the location list is found by a regular search in the request, and that the request contains a sum() and groupbys, it could be done with OpenERP but would be much slower for 2 non-stored function fields."

lepistone commented:
I'm merging this. Thanks to all!
